pallets/delegation/src/delegation_hierarchy.rs

```diff
@@ -52,87 +57,158 @@ impl Default for Permissions {
-	/// perform.
-	pub permissions: Permissions,
-	/// The flag indicating whether the delegation has been revoked or not.
-	pub revoked: bool,
+	/// A node representing a node in the delegation hierarchy.
+	#[derive(Clone, Debug, Encode, Decode, PartialEq)]
+	pub struct DelegationNode<T: Config> {
+		/// The ID of the delegation hierarchy root.
+		pub root_id: DelegationNodeIdOf<T>,
+		/// \[OPTIONAL\] The ID of the parent node. If None, the node is
+		/// considered a direct child of the root node.
+		pub parent: Option<DelegationNodeIdOf<T>>,
+		/// The identifier of the owner of the delegation node, i.e., the
+		/// delegate.
+		pub owner: DelegatorIdOf<T>,
+		/// The permission flags for the operations the delegate is allowed to
+		/// perform.
+		pub permissions: Permissions,
+		/// The flag indicating whether the delegation has been revoked or not.
+		pub revoked: bool,
+	}
+
+	impl<T: Config> DelegationNode<T> {
+		/// Create a new delegation node that is a direct descendent of the
+		/// given root.
+		///
+		/// * root_id: the root node ID this node will be a child of
+		/// * owner: the identifier of the owner of the new delegation, i.e.,
+		///   the new delegate
+		/// * permissions: the permission flags for the operations the delegate
+		///   is allowed to perform
+		pub fn new_root_child(
+			root_id: DelegationNodeIdOf<T>,
+			owner: DelegatorIdOf<T>,
+			permissions: Permissions,
+		) -> Self {
+			DelegationNode {
+				root_id,
+				owner,
+				permissions,
+				revoked: false,
+				parent: None,
+			}
+		}
+
+		/// Creates a new delegation node that is a direct descendent of the
+		/// given node.
+		///
+		/// * root_id: the root node ID this node will be a child of
+		/// * parent - the parent node ID this node will be a child of
+		/// * owner: the identifier of the owner of the new delegation, i.e.,
+		///   the new delegate
+		/// * permissions: the permission flags for the operations the delegate
+		///   is allowed to perform
+		pub fn new_node_child(
+			root_id: DelegationNodeIdOf<T>,
+			parent: DelegationNodeIdOf<T>,
+			owner: DelegatorIdOf<T>,
+			permissions: Permissions,
+		) -> Self {
+			DelegationNode {
+				root_id,
+				parent: Some(parent),
+				owner,
+				permissions,
+				revoked: false,
+			}
+		}
+	}
 }
 
-impl<T: Config> DelegationNode<T> {
-	/// Create a new delegation node that is a direct descendent of the
-	/// given root.
-	///
-	/// * root_id: the root node ID this node will be a child of
-	/// * owner: the identifier of the owner of the new delegation, i.e., the
-	///   new delegate
-	/// * permissions: the permission flags for the operations the delegate is
-	///   allowed to perform
-	pub fn new_root_child(root_id: DelegationNodeIdOf<T>, owner: DelegatorIdOf<T>, permissions: Permissions) -> Self {
-		DelegationNode {
-			root_id,
-			owner,
-			permissions,
-			revoked: false,
-			parent: None,
+// The new types.
+pub(crate) mod v1 {
```

Comment:
When the old types are in their own file i would just have the current types in the root? I think that's just a bit more straight forward. You don't have to think about which version is the latest etc.

Reply:
Addressed in https://github.com/KILTprotocol/mashnet-node/commit/0040f2d75ca135d86b18c86cea6f54e0cb7e24e5.